pkg/storage/provider_abort_action_test.go

```diff
@@ -0,0 +1,296 @@
+package storage_test
+
+import (
+	"testing"
+
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/fixtures"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/fixtures/testusers"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/storage/internal/testabilities"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/wdk"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAbortActionSuccess(t *testing.T) {
+	// given:
+	given, cleanup := testabilities.Given(t)
+	defer cleanup()
+
+	activeStorage := given.Provider().GORM()
+	createResult, _ := given.ActionCreatedAndSigned(activeStorage)
+
+	// when:
+	result, err := activeStorage.AbortAction(
+		t.Context(),
+		testusers.Alice.AuthID(),
+		wdk.AbortActionArgs{
+			Reference: createResult.Reference,
+		},
+	)
+
+	// then:
+	require.NoError(t, err)
+	require.NotNil(t, result)
+	require.True(t, result.Aborted)
+
+	thenDBState := testabilities.ThenDBState(t, activeStorage)
+	thenDBState.HasUserTransactionByReference(testusers.Alice, createResult.Reference).
+		WithStatus(wdk.TxStatusFailed)
+	thenDBState.AllOutputs(testusers.Alice).WithCount(1)
+}
+
+func TestAbortActionSuccessfulSpendingAfterAbort(t *testing.T) {
+	// given:
+	given, cleanup := testabilities.Given(t)
+	defer cleanup()
+
+	activeStorage := given.Provider().GORM()
+	createResult, _ := given.ActionCreatedAndSigned(activeStorage)
+
+	// when:
+	result, err := activeStorage.AbortAction(
+		t.Context(),
+		testusers.Alice.AuthID(),
+		wdk.AbortActionArgs{
+			Reference: createResult.Reference,
+		},
+	)
+	require.NoError(t, err)
+
+	createResult, err = activeStorage.CreateAction(
+		t.Context(),
+		testusers.Alice.AuthID(),
+		fixtures.DefaultValidCreateActionArgs(),
+	)
+
+	// then:
+	require.NoError(t, err)
+	require.NotNil(t, result)
+	require.True(t, result.Aborted)
+
+	thenDBState := testabilities.ThenDBState(t, activeStorage)
+	thenDBState.HasUserTransactionByReference(testusers.Alice, createResult.Reference).
+		WithStatus(wdk.TxStatusUnsigned)
+}
+
+func TestAbortActionErrorCases(t *testing.T) {
+	tests := map[string]struct {
+		setupTransaction func(t *testing.T, given testabilities.StorageFixture) (string, wdk.AuthID)
+		args             func(reference string) wdk.AbortActionArgs
+		expectedErrors   []string
+	}{
+		"transaction not found by reference": {
+			setupTransaction: func(t *testing.T, given testabilities.StorageFixture) (string, wdk.AuthID) {
+				return "non-existent-reference", testusers.Alice.AuthID()
+			},
+			args: func(reference string) wdk.AbortActionArgs {
```

Comment:
I'm almost sure that all those cases have the same implementation of args function